### DISCUSSION TITLE
Using validation set for early stopping

### DISCUSSION THREAD

**Discussion by sivomke** [ContinualAI/avalanche#900] [Q&A]
Hi, everyone!

I have created a benchmark with **dataset_benchmark** that contains 3 experiences, and then have added a validation set to this benchmark with **benchmark_with_validation_stream**. I am trying to use Early Stopping with validation set: for each experience, I would like to use loss on the validation set of the corresponding experience. I have indicated **metric_name = 'Loss_Exp'** and  **val_stream_name='valid'** , however, it still uses **Loss_Exp/eval_phase/train_stream** of the corresponding experience instead.  Could you please tell me what would be the correct way to use validation set for early stopping?

[Here is the script](https://github.com/sivomke/avalanche/blob/main/base-strategy.py)

Thank you!

**Comment by vlomonaco** [comment]
Hi @sivomke! I had a quick look into your script and I didn't find any issue. This may be a bug. We will look into it asap!

**Reply by sivomke** [reply]
Thank you!

**Reply by sivomke** [reply]
Adding corresponding validation experience to train function seems to solve the problem with using the correct stream:
```
for i, experience in enumerate(benchmark.train_stream):
    print("Start of experience: ", experience.current_experience)
    res = strategy.train(experience, eval_streams = [benchmark.valid_stream[i]])
    print('Training completed')
```
It starts evaluating on the **Loss_Exp/eval_phase/valid_stream/..**, however, it does not stop training when the loss on validation experience has been increasing for **patience** number of epochs.

**Reply by vlomonaco** [reply]
Thanks for this additional info @sivomke! Can you send us a minimal example to reproduce the issue? Thanks!

**Reply by AntonioCarta** [reply]
Can you try to set `eval_every=1` in your strategy? The `EarlyStoppingPlugin` only checks the last metric value. However, the metric is updated only if the evaluation is called after the epochs, which is controlled by `eval_every`. If you set `eval_every=10` it means that you only evaluate every 10 epochs.

**Reply by sivomke** [reply]
I have set `eval_every` and `patience` to be the same value, since I probably have misunderstood the comment in `EarlyStoppingPlugin` definition:


        For example, if you set `patience=1`, you must also set `eval_every=1`
        in the `BaseStrategy`, otherwise the metric won't be updated after
        every epoch/iteration. Similarly, `peval_mode` must have the same
        value.

I have set `eval_every=1` in the strategy, and also added printing statement to  `before_training_epoch` function to track the issue. And it turns out that it does not stop since the `self.best_val` is not being updated, it seems to be because `self.metric_key` does not correspond to the full name of the metric that is being computed, since it also contain Task name and Experience name:

- if we set `EarlyStoppingPlugin(patience=2, metric_name = 'Loss_Exp', val_stream_name='valid', mode='min', peval_mode='epoch')` , we have:
![image](https://user-images.githubusercontent.com/32356440/152313411-1259cbc9-ad6c-4c72-b282-5411c55f2d9a.png)
- same if we set `EarlyStoppingPlugin(patience=2, metric_name = 'Loss_Exp', val_stream_name='valid_stream', mode='min', peval_mode='epoch')` , we have:
![image](https://user-images.githubusercontent.com/32356440/152313489-f75ae1c6-2fd4-4336-8f1c-782f05d43c89.png)

I will update the [current script ](https://github.com/sivomke/avalanche/blob/main/base-strategy.py) with a dataset for it to be a minimal example to reproduce the issue a bit later since I cannot upload the data I am currently using.

Thank you for your time and attention!

**Reply by AntonioCarta** [reply]
Thanks. This looks like a bug. If you open an issue with a minimal example it can help to fix it quickly.

**Reply by sivomke** [reply]
I have created [minimal example](https://github.com/sivomke/avalanche/blob/main/min-reproducable-example.py):
`python min-reproducable-example.py  --early_stopping --training_batch_size 8 --eval_batch_size 1 --epochs 25 --lr 2e-4 `
For first experience, we have the first 75 samples with target 'zeros', and last 25 samples with target 'ones', in `benchmark_with_validation_stream` we create 25% validation split, and by setting `shuffle=False` we use the last 25 samples of the first experience as a validation set, hence for the first experience the error on validation set will increase.
![image](https://user-images.githubusercontent.com/32356440/152336924-abd863a9-04ab-4ac0-883d-2e71b41a1f1d.png)

**Reply by vlomonaco** [reply]
I created #904 to track the issue!

**Reply by sivomke** [reply]
Hello!

I have found a [workaround](https://github.com/sivomke/avalanche/blob/main/min-reproducable-example-potential-fix.py) that works for my application, however, it may not be the most general way to do this. Since the problem was that the full name of the metric also contains the experience, and for each experience, I evaluate on the validation set of the corresponding experience (`res = strategy.train(experience, eval_streams=[benchmark.valid_stream[i]])`), instead of trying to retrieve the metric by the key (in the example `Loss_Exp/eval_phase/valid_stream`, which caused the problem since the full name of the computed metric contains the experience and task name), I retrieve the last computed metric that starts with the key :
```
res = strategy.evaluator.get_last_metrics()
full_name =[k for k in res.keys() if k.startswith(self.metric_key)][-1]
val_acc = res.get(full_name)
```
Hence in my application, during the training on the first experience, the retrieved value would be `Loss_Exp/eval_phase/valid_stream/Task000/Exp000`, while for the second experience it would be `Loss_Exp/eval_phase/valid_stream/Task000/Exp001`.

**Reply by vlomonaco** [reply]
Thanks @sivomke for reporting your fix (that may be useful to others) and for your patience, we will try to solve the bug soon.
